src/group_theory/perm/cycle_type.lean

```diff
@@ -0,0 +1,403 @@
+  { intros σ τ hστ hσ hτ,
+    rw [hστ.cycle_type, sum_add, hστ.card_support_mul, hσ, hτ] },
+end
+
+lemma lcm_cycle_type (σ : perm α) : σ.cycle_type.lcm = order_of σ :=
+cycle_induction_on
+  (λ τ : perm α, τ.cycle_type.lcm = order_of τ) σ
+  (by rw [cycle_type_one, lcm_zero, order_of_one])
+  (λ σ hσ, by rw [hσ.cycle_type, ←singleton_coe, lcm_singleton, order_of_is_cycle hσ,
+    nat.normalize_eq])
+  (λ σ τ hστ hσ hτ, by rw [hστ.cycle_type, lcm_add, nat.lcm_eq_lcm, hστ.order_of, hσ, hτ])
+
+lemma dvd_of_mem_cycle_type {σ : perm α} {n : ℕ} (h : n ∈ σ.cycle_type) : n ∣ order_of σ :=
+begin
+  rw ← lcm_cycle_type,
+  exact dvd_lcm h,
+end
+
+lemma cycle_type_prime_order {σ : perm α} (hσ : (order_of σ).prime) :
+  ∃ n : ℕ, σ.cycle_type = repeat (order_of σ) n.succ :=
+begin
+  rw eq_repeat_of_mem (λ n hn, or_iff_not_imp_left.mp
+    (hσ.2 n (dvd_of_mem_cycle_type hn)) (ne_of_gt (one_lt_of_mem_cycle_type hn))),
+  use σ.cycle_type.card.pred,
+  rw nat.succ_pred_eq_of_pos,
+  rw [pos_iff_ne_zero, ne, card_cycle_type_eq_zero],
+  rintro rfl,
+  rw order_of_one at hσ,
+  exact hσ.ne_one rfl,
+end
+
+lemma is_cycle_of_prime_order {σ : perm α} (h1 : (order_of σ).prime)
+  (h2 : σ.support.card < 2 * (order_of σ)) : σ.is_cycle :=
+begin
+  obtain ⟨n, hn⟩ := cycle_type_prime_order h1,
+  rw [←σ.sum_cycle_type, hn, multiset.sum_repeat, nsmul_eq_mul, nat.cast_id, mul_lt_mul_right
+      (order_of_pos σ), nat.succ_lt_succ_iff, nat.lt_succ_iff, nat.le_zero_iff] at h2,
+  rw [←card_cycle_type_eq_one, hn, card_repeat, h2],
+end
+
+end cycle_type
+
+lemma is_cycle_of_prime_order' {σ : perm α} (h1 : (order_of σ).prime)
+  (h2 : fintype.card α < 2 * (order_of σ)) : σ.is_cycle :=
+begin
+  classical,
+  exact is_cycle_of_prime_order h1 (lt_of_le_of_lt σ.support.card_le_univ h2),
+end
+
+lemma is_cycle_of_prime_order'' {σ : perm α} (h1 : (fintype.card α).prime)
+  (h2 : order_of σ = fintype.card α) : σ.is_cycle :=
+is_cycle_of_prime_order' ((congr_arg nat.prime h2).mpr h1)
+begin
+  classical,
+  rw [←one_mul (fintype.card α), ←h2, mul_lt_mul_right (order_of_pos σ)],
+  exact one_lt_two,
+end
+
+section partition
+
+variables [decidable_eq α]
+
+/-- The partition corresponding to a permutation -/
+def partition (σ : perm α) : partition (fintype.card α) :=
+{ parts := σ.cycle_type + repeat 1 (fintype.card α - σ.support.card),
+  parts_pos := λ n hn,
+  begin
+    cases mem_add.mp hn with hn hn,
+    { exact zero_lt_one.trans (one_lt_of_mem_cycle_type hn) },
+    { exact lt_of_lt_of_le zero_lt_one (ge_of_eq (multiset.eq_of_mem_repeat hn)) },
+  end,
+  parts_sum := by rw [sum_add, sum_cycle_type, multiset.sum_repeat, nsmul_eq_mul,
+    nat.cast_id, mul_one, nat.add_sub_cancel' σ.support.card_le_univ] }
+
+end partition
+
+/-!
+### 3-cycles
+-/
+
+section is_three_cycle
+
+variables [decidable_eq α] {σ : perm α}
+
+/-- A three-cycle is a cycle of length 3. -/
+def is_three_cycle (σ : perm α) : Prop := σ.support.card = 3
```

Comment:
I fiddled around with defining `is_three_cycle` in terms of `cycle_type`, and here's what I got. It doesn't really shorten things though.
```suggestion
/-- A three-cycle is a cycle of length 3. -/
def is_three_cycle (σ : perm α) : Prop := σ.cycle_type = {3}

lemma is_three_cycle.cycle_type (h : is_three_cycle σ) : σ.cycle_type = {3} := h

lemma is_three_cycle.card_support (h : is_three_cycle σ) : σ.support.card = 3 :=
by rw [←sum_cycle_type, h.cycle_type, singleton_eq_singleton, multiset.sum_cons, sum_zero]

lemma card_support_eq_three_iff : σ.support.card = 3 ↔ σ.is_three_cycle :=
begin
  refine ⟨λ h, _, is_three_cycle.card_support⟩,
  by_cases h0 : σ.cycle_type.card = 0,
  { rw [←sum_cycle_type, card_eq_zero.mp h0, sum_zero] at h,
    exact (ne_of_lt zero_lt_three h).elim },
  by_cases h1 : σ.cycle_type.card = 1,
  { obtain ⟨n, hn⟩ := card_eq_one.mp h1,
    rw [←sum_cycle_type, hn, multiset.sum_cons, sum_zero, add_zero] at h,
    rw [is_three_cycle, hn, h, singleton_eq_singleton] },
  { have hle := card_nsmul_le_sum (λ x, two_le_of_mem_cycle_type),
    rw [nat.nsmul_eq_mul, sum_cycle_type, h, ←nat.lt_succ_iff] at hle,
    exact (ne_of_lt (lt_of_le_of_lt (nat.mul_le_mul_right 2 (nat.succ_le_iff.mpr
      (nat.one_lt_iff_ne_zero_and_ne_one.mpr ⟨h0, h1⟩))) hle) rfl).elim },
end

lemma is_three_cycle.is_cycle (h : is_three_cycle σ) : is_cycle σ :=
by rw [←card_cycle_type_eq_one, h.cycle_type, singleton_eq_singleton, card_singleton]

lemma is_three_cycle.sign (h : is_three_cycle σ) : sign σ = 1 :=
begin
  rw [sign_of_cycle_type, h.cycle_type],
  refl,
end

lemma is_three_cycle.inv {f : perm α} (h : is_three_cycle f) : is_three_cycle (f⁻¹) :=
by rwa [is_three_cycle, cycle_type_inv]

lemma is_three_cycle_swap_mul_swap_same {a b c : α} (ab : a ≠ b) (ac : a ≠ c) (bc : b ≠ c) :
  is_three_cycle (swap a b * swap a c) :=
begin
  suffices h : support (swap a b * swap a c) = {a, b, c},
  { rw [←card_support_eq_three_iff, h],
    simp [ab, ac, bc] },
  apply le_antisymm ((support_mul_le _ _).trans (λ x, _)) (λ x hx, _),
  { simp [ab, ac, bc] },
  { simp only [finset.mem_insert, finset.mem_singleton] at hx,
    rw mem_support,
    simp only [perm.coe_mul, function.comp_app, ne.def],
    obtain rfl | rfl | rfl := hx,
    { rw [swap_apply_left, swap_apply_of_ne_of_ne ac.symm bc.symm],
      exact ac.symm },
    { rw [swap_apply_of_ne_of_ne ab.symm bc, swap_apply_right],
      exact ab },
    { rw [swap_apply_right, swap_apply_left],
      exact bc } }
end
```

Reply:
This is great, thanks! Now to see if I can get rid of the `multiset` part of the diff...